### ISSUE TITLE
Delay bugged. :(

### ISSUE THREAD

**Issue by heniotierra** [craftyjs/Crafty#761]
Hello! I noticed Delay has a problem, problem that was present prior to the modifications I made recently. In fact, I had noticed this problem some months ago, but did circumvent it with an ugly hack at the time. What happens is that the delay acts in a weird manner when there's more than one simultaneous delay set in the same entity, making the same callback be called twice (or thrice) when it should be called only once. I tried really hard to figure out why this happens, but I was unable to.

The function .cancelDelay that I made makes the problem more obvious. My only guess is that when a delay is removed with splice, it messes with the logic of the loop in some way.

That said, I found a way to fix the problem, setting the given delay as undefined instead of using .splice to remove it from the array. Then, I used another loop just to use splice, and that worked pretty well.

 EDIT: Here's the fiddle showing the bug: http://jsfiddle.net/Dj48n/9/. Also, I'll do a new PR with my fix for that.

**Comment by mucaho** [comment]
I just made another PR adding some tests #760 , so I could include the fix in there, once we find the bug :)

Good that you catched that: the current testcases do not test setting up multiple delays, I will add those testcases now.

One thing that boggled my mind is that `Delay` uses property enumeration on the `_delays` array instead of array iteration. Initially I thought it doesn't matter, as the functionality does not depend on the order of iteration, but maybe it does matter after all.

**Comment by mucaho** [comment]
I have added additional testcases. If there is a bug then I think it's local to the cancelDelay method:

``` javascript
cancelDelay: function (callback) {
    for (var index in this._delays) {
        var item = this._delays[index];
        if(item.callback == callback) {
            this._delays.splice(index, 1);
        }
    }
    return this;
}
```

First of all index will be randomly selected in the for loop between `0` and `this._delays.length - 1`.
When you splice an element out of the array, any old indices won't be valid any more (indices coming after the spliced element have to be shifted `1` to the left). So I am assuming the iteration over the `indices` is erroneous.

**Comment by mucaho** [comment]
Yeah, that was the issue. Adding the fix to #760 .

**Comment by mucaho** [comment]
I merged #760, so I will assume the bug is fixed. If you encounter another bug related to `Delay`, just reopen the issue.

**Comment by heniotierra** [comment]
Unfortunately that didn't solve the bug I refer to. Here's the fiddle: http://jsfiddle.net/Dj48n/9/. One alert should appear at the end. 

Shall I do a PR with my fix?

**Comment by mucaho** [comment]
Sure, that would be great!

**Comment by heniotierra** [comment]
Alright, just let me do these grunt and git stuff. :+1:

**Comment by mucaho** [comment]
In your jsfiddle on L167 you schedule 3 `createBigBox` delays.
Each of this delays schedules one anonymous delay.
The anonymous delay schedules an `createBigBox` delay.
The `createBigBox` delay schedules an _alert_.
=> Could that be the reason why 3 times _alert_ is called?

`cancelDelay` is only called for `createSmallBox`

**Comment by heniotierra** [comment]
'this' refers to the entity with Delay component, right? So, it's the same delay entity.

I only canceled the createSmallBox delays because they are the only ones that I want to cancel, for they receive -1 as repeat.

There should be only 1 big box in the end, but 3 are created too.

**Comment by heniotierra** [comment]
@mucaho There you go... https://github.com/craftyjs/Crafty/issues/762 (this github thing is some smart shit)

PS.: The second loop is not necessary if you do not care about having the _delays array with null values.

PS.²: Here you can see the fiddle with the fix from my PR. http://jsfiddle.net/Dj48n/14/

**Comment by mucaho** [comment]
>  My only guess is that when a delay is cancelled with splice, it messes with the logic of the loop in some way.

Yep that's it!

**Comment by mucaho** [comment]
Fixed with #762 , closing.
